CoreScriptsRoot/Modules/BackpackScript3D.lua

```diff
@@ -0,0 +1,259 @@
+local ICON_SIZE = 48
+local ICON_SPACING = 52
+local PIXELS_PER_STUD = 64
+
+local HOPPERBIN_ANGLE = math.rad(-45)
+local HOPPERBIN_ROTATION = CFrame.Angles(HOPPERBIN_ANGLE, 0, 0)
+local HOPPERBIN_OFFSET = Vector3.new(0, 0, -5)
+
+local Tools = {}
+local ToolsList = {}
+
+local BackpackScript = {}
+local topbarEnabled = false
+
+local player = game.Players.LocalPlayer
+
+local hopperbinPart = Instance.new("Part", workspace.CurrentCamera)
+hopperbinPart.Transparency = 1
+hopperbinPart.CanCollide = false
+hopperbinPart.Anchored = true
+hopperbinPart.Name = "GUI"
+local hopperbinGUI = Instance.new("SurfaceGui", player:WaitForChild("PlayerGui"))
+hopperbinGUI.Adornee = hopperbinPart
+hopperbinGUI.ToolPunchThroughDistance = 1000
+hopperbinGUI.Name = "HopperBinGUI"
+
+local verticalRange = math.rad(0)
+local horizontalRange = math.rad(0)
+
+local function DoLayout()
+	local width, height = 100, 100
+	local borderSize = (ICON_SPACING - ICON_SIZE) / 2	
+	
+	local x = borderSize
+	for i, tool in ipairs(ToolsList) do
```

Comment:
I did this to keep the order of ToolsList while also maintaining a lookup table for metadata on each tool (the ImageButton, methods with specific upvalues, etc.)  I used ipairs because it's slightly faster than pairs in my experience.